File: dertetgpt-ios/DertetHarness/Network/AnthropicClient.swift
import Foundation

final class AnthropicClient: ChatClient {
    func streamChat(
        baseUrl: String,
        apiKey: String,
        model: String,
        systemPrompt: String,
        history: [ChatMessage]
    ) -> AsyncThrowingStream<ChatStreamEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                var gotAny = false
                do {
                    var messages: [[String: Any]] = []
                    for msg in history {
                        messages.append(["role": msg.role, "content": buildContent(msg)])
                    }

                    var body: [String: Any] = [
                        "model": model,
                        "max_tokens": 8192,
                        "stream": true,
                        "messages": messages
                    ]
                    if !systemPrompt.isEmpty { body["system"] = systemPrompt }

                    guard let url = URL(string: baseUrl.trimmedTrailingSlash + "/messages") else {
                        continuation.finish(throwing: URLError(.badURL))
                        return
                    }
                    var request = URLRequest(url: url)
                    request.httpMethod = "POST"
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
                    request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
                    request.httpBody = jsonString(body)

                    for try await event in sseEvents(for: request) {
                        try Task.checkCancellation()
                        let data = event.data.trimmingCharacters(in: .whitespaces)
                        if data.isEmpty { continue }
                        guard let jsonData = data.data(using: .utf8),
                              let obj = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else { continue }

                        switch event.type {
                        case "content_block_delta":
                            if let delta = asDict(obj["delta"]), let text = asString(delta["text"]), !text.isEmpty {
                                gotAny = true
                                continuation.yield(.delta(text))
                            }
                        case "error":
                            if let errorObj = asDict(obj["error"]), let msg = asString(errorObj["message"]) {
                                continuation.yield(.error(msg))
                            }
                        default:
                            break
                        }
                    }
                    continuation.yield(.done)
                    continuation.finish()
                } catch {
                    if !gotAny {
                        continuation.yield(.error(error.localizedDescription))
                    }
                    continuation.finish()
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    private func buildContent(_ msg: ChatMessage) -> Any {
        let images = msg.attachments.filter { $0.kind == .image && $0.base64Data != nil }
        let files = msg.attachments.filter { $0.kind == .file }

        var text = msg.text
        for f in files where f.textContent != nil {
            text = "[File: \(f.fileName)]\n\(f.textContent!)\n\n\(text)"
        }

        if images.isEmpty { return text }

        var parts: [[String: Any]] = []
        for img in images {
            parts.append([
                "type": "image",
                "source": ["type": "base64", "media_type": img.mimeType, "data": img.base64Data!]
            ])
        }
        parts.append(["type": "text", "text": text])
        return parts
    }
}

File: dertet-harness-android/app/src/main/java/com/dertet/harness/network/AnthropicClient.kt
package com.dertet.harness.network

import com.dertet.harness.data.AttachmentKind
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class AnthropicClient : ChatClient {

    override fun streamChat(
        baseUrl: String,
        apiKey: String,
        model: String,
        systemPrompt: String,
        history: List<ChatMessage>
    ): Flow<ChatStreamEvent> = flow {
        val messages = buildJsonArray {
            for (msg in history) {
                add(buildJsonObject {
                    put("role", msg.role)
                    put("content", buildContent(msg))
                })
            }
        }

        val bodyBuilder = buildJsonObject {
            put("model", model)
            put("max_tokens", 8192)
            put("stream", true)
            if (systemPrompt.isNotBlank()) put("system", systemPrompt)
            put("messages", messages)
        }
        val body = bodyBuilder.toString().toRequestBody("application/json".toMediaType())

        val url = baseUrl.trimEnd('/') + "/messages"
        val request = Request.Builder()
            .url(url)
            .addHeader("x-api-key", apiKey)
            .addHeader("anthropic-version", "2023-06-01")
            .addHeader("Content-Type", "application/json")
            .post(body)
            .build()

        var gotAny = false
        try {
            sseFlow(sharedHttpClient, request).collect { event ->
                val type = event.type
                val data = event.data.trim()
                if (data.isEmpty()) return@collect
                try {
                    when (type) {
                        "content_block_delta" -> {
                            val json = Json.parseToJsonElement(data).jsonObject
                            val delta = json["delta"]?.jsonObject
                            val text = delta?.get("text")?.jsonPrimitive?.contentOrNull
                            if (!text.isNullOrEmpty()) {
                                gotAny = true
                                emit(ChatStreamEvent.Delta(text))
                            }
                        }
                        "error" -> {
                            val json = Json.parseToJsonElement(data).jsonObject
                            val msg = json["error"]?.jsonObject?.get("message")?.jsonPrimitive?.contentOrNull
                                ?: "API error"
                            emit(ChatStreamEvent.Error(msg))
                        }
                        else -> {
                            // message_start, content_block_start/stop, message_delta, message_stop, ping - ignored
                        }
                    }
                } catch (e: Exception) {
                    // ignore malformed chunk
                }
            }
            emit(ChatStreamEvent.Done)
        } catch (e: Exception) {
            if (!gotAny) emit(ChatStreamEvent.Error(e.message ?: "Помилка з’єднання"))
            else emit(ChatStreamEvent.Done)
        }
    }

    private fun buildContent(msg: ChatMessage): kotlinx.serialization.json.JsonElement {
        val images = msg.attachments.filter { it.kind == AttachmentKind.IMAGE && it.base64Data != null }
        val files = msg.attachments.filter { it.kind == AttachmentKind.FILE }

        var textPart = msg.text
        for (f in files) {
            if (f.textContent != null) {
                textPart = "[Файл: ${f.fileName}]\n${f.textContent}\n\n$textPart"
            }
        }

        if (images.isEmpty()) {
            return JsonPrimitive(textPart)
        }

        return buildJsonArray {
            for (img in images) {
                add(buildJsonObject {
                    put("type", "image")
                    put("source", buildJsonObject {
                        put("type", "base64")
                        put("media_type", img.mimeType)
                        put("data", img.base64Data)
                    })
                })
            }
            add(buildJsonObject {
                put("type", "text")
                put("text", textPart)
            })
        }
    }
}
